File: sdk/swift/Sources/Plinth/KV.swift
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 The Plinth Authors

import Foundation

/// Versioned key-value store for a workspace.
///
/// Every ``set(key:value:)`` creates a new immutable version. Reads
/// default to the latest version; pass an explicit ``version`` to read
/// a specific historical revision.
///
/// Construct via ``Workspace/kv``.
public struct KV: Sendable {
    let http: HTTPClient
    let workspaceID: String

    init(http: HTTPClient, workspaceID: String) {
        self.http = http
        self.workspaceID = workspaceID
    }

    /// Write `value` to `key` and return the resulting versioned entry.
    ///
    /// `value` is JSON-encoded — pass any `Encodable` (including
    /// ``AnyCodableValue``) or a Foundation primitive.
    @discardableResult
    public func set<V: Encodable>(key: String, value: V) async throws -> KVEntry {
        let body = SetRequest(value: value)
        return try await http.putJSON(
            path(for: key),
            body: body,
            notFoundCode: PlinthErrorCode.workspaceNotFound
        )
    }

    /// Convenience: set an arbitrary Foundation value (Bool, Int, Double,
    /// String, Array, Dictionary).
    @discardableResult
    public func setAny(key: String, value: Any?) async throws -> KVEntry {
        return try await set(key: key, value: AnyCodableValue.from(value))
    }

    /// Read the latest value for `key`, decoded as `V`.
    ///
    /// Throws ``PlinthError/keyNotFound`` when the key was deleted or
    /// never written.
    public func get<V: Decodable>(key: String, as type: V.Type = V.self) async throws -> V {
        let entry: KVEntry = try await getEntry(key: key)
        // The wire value is wrapped in AnyCodableValue; re-encode then
        // decode into the user's target type for a transparent cast.
        let raw = try http.encoder.encode(entry.value)
        do {
            return try http.decoder.decode(V.self, from: raw)
        } catch {
            throw PlinthError.decoding(error.localizedDescription)
        }
    }

    /// Read the latest entry (value + metadata) for `key`.
    public func getEntry(key: String) async throws -> KVEntry {
        return try await http.getJSON(
            path(for: key),
            notFoundCode: PlinthErrorCode.keyNotFound
        )
    }

    /// Read a specific historical version of `key`.
    public func getVersion(key: String, version: Int) async throws -> KVEntry {
        return try await http.getJSON(
            path(for: key),
            query: ["version": String(version)],
            notFoundCode: PlinthErrorCode.keyNotFound
        )
    }

    /// Every recorded version of `key`, oldest first.
    public func history(key: String) async throws -> [KVEntry] {
        let resp: KVHistoryResponse = try await http.getJSON(
            "\(path(for: key))/history",
            notFoundCode: PlinthErrorCode.keyNotFound
        )
        return resp.versions
    }

    /// Tombstone `key`. Reads after this throw ``PlinthError/keyNotFound``.
    public func delete(key: String) async throws {
        try await http.delete(
            path(for: key),
            notFoundCode: PlinthErrorCode.keyNotFound
        )
    }

    /// Latest entry for every key in the workspace.
    public func list() async throws -> [KVEntry] {
        let resp: KVListResponse = try await http.getJSON(
            "/v1/workspaces/\(encodePathSegment(workspaceID))/kv",
            notFoundCode: PlinthErrorCode.workspaceNotFound
        )
        return resp.entries
    }

    // MARK: - Internal helpers

    func path(for key: String) -> String {
        return "/v1/workspaces/\(encodePathSegment(workspaceID))/kv/\(encodePathSegment(key))"
    }

    /// Request body wrapper for PUT /v1/workspaces/{id}/kv/{key}.
    struct SetRequest<V: Encodable>: Encodable {
        let value: V
    }
}

File: sdk/kotlin/src/main/kotlin/dev/plinth/sdk/KV.kt
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 The Plinth Authors

package dev.plinth.sdk

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement

/**
 * Versioned key-value store for a workspace.
 *
 * Every [set] creates a new immutable version. Reads default to the
 * latest version; pass an explicit version to [getVersion] for a
 * specific historical revision.
 *
 * Construct via [Workspace.kv].
 */
class KV internal constructor(
    @PublishedApi internal val http: HttpClient,
    private val workspaceId: String,
) {
    @PublishedApi internal val json: Json get() = http.json

    /**
     * Write [value] to [key] and return the versioned entry.
     *
     * `value` can be any JSON-serializable Kotlin type — strings, ints,
     * lists, maps, or your own `@Serializable` data classes. Pass a
     * [JsonElement] directly when you've already built one.
     */
    suspend inline fun <reified T> set(key: String, value: T): KVEntry =
        setSerialized(key, json.encodeToJsonElement(value))

    /**
     * Write a pre-built [JsonElement] to [key] without going through
     * the inline reified path. Useful when the serializer is computed
     * at runtime.
     */
    suspend fun setSerialized(key: String, value: JsonElement): KVEntry {
        return http.putJson(
            path = path(key),
            body = KVSetRequest(value),
            bodySerializer = KVSetRequest.serializer(),
            responseSerializer = KVEntry.serializer(),
            notFoundCode = PlinthErrorCode.WORKSPACE_NOT_FOUND,
        )
    }

    /**
     * Read the latest value for [key] decoded as [T].
     *
     * Throws [PlinthError.KeyNotFound] when the key was deleted or
     * never written.
     */
    suspend inline fun <reified T> get(key: String): T {
        val entry = getEntry(key)
        return json.decodeFromJsonElement<T>(entry.value)
    }

    /**
     * Read the latest entry (value + metadata) for [key].
     */
    suspend fun getEntry(key: String): KVEntry =
        http.getJson(
            path = path(key),
            deserializer = KVEntry.serializer(),
            notFoundCode = PlinthErrorCode.KEY_NOT_FOUND,
        )

    /**
     * Read a specific historical [version] of [key].
     */
    suspend fun getVersion(key: String, version: Int): KVEntry =
        http.getJson(
            path = path(key),
            deserializer = KVEntry.serializer(),
            query = mapOf("version" to version.toString()),
            notFoundCode = PlinthErrorCode.KEY_NOT_FOUND,
        )

    /** Every recorded version of [key], oldest first. */
    suspend fun history(key: String): List<KVEntry> =
        http.getJson(
            path = "${path(key)}/history",
            deserializer = KVHistoryResponse.serializer(),
            notFoundCode = PlinthErrorCode.KEY_NOT_FOUND,
        ).versions

    /** Tombstone [key]. Reads after this throw [PlinthError.KeyNotFound]. */
    suspend fun delete(key: String) {
        http.delete(path(key), notFoundCode = PlinthErrorCode.KEY_NOT_FOUND)
    }

    /** Latest entry for every key in the workspace. */
    suspend fun list(): List<KVEntry> =
        http.getJson(
            path = "/v1/workspaces/${encodePathSegment(workspaceId)}/kv",
            deserializer = KVListResponse.serializer(),
            notFoundCode = PlinthErrorCode.WORKSPACE_NOT_FOUND,
        ).entries

    // MARK: - Helpers

    private fun path(key: String): String =
        "/v1/workspaces/${encodePathSegment(workspaceId)}/kv/${encodePathSegment(key)}"
}
